File: apps/api_python/src/class_reflect_api/application/agent_orchestrator.py
from dataclasses import dataclass
from enum import StrEnum


class AgentType(StrEnum):
    TRANSCRIPTION = "transcription_agent"
    TRANSLATION = "translation_agent"
    EVIDENCE = "evidence_agent"
    REPORT = "report_agent"
    TEACHER_REVIEW = "teacher_review_agent"


class OrchestratorAction(StrEnum):
    EXECUTE = "execute"
    ASK_TEACHER = "ask_teacher"
    RETRY_OR_WAIT = "retry_or_wait_for_teacher"


@dataclass(frozen=True)
class OrchestratorDecision:
    action: OrchestratorAction
    next_agent: AgentType
    reason: str


def decide_next_action(
    *,
    task_status: str | None = None,
    workflow_status: str | None = None,
    requires_translation: bool = False,
) -> OrchestratorDecision:
    if task_status == "failed" or workflow_status == "failed":
        return OrchestratorDecision(
            action=OrchestratorAction.RETRY_OR_WAIT,
            next_agent=AgentType.TEACHER_REVIEW,
            reason="当前流程失败，需要教师确认后重试或调整条件。",
        )

    if requires_translation:
        return OrchestratorDecision(
            action=OrchestratorAction.EXECUTE,
            next_agent=AgentType.TRANSLATION,
            reason="教师主动要求翻译，进入可选翻译 Agent。",
        )

    return OrchestratorDecision(
        action=OrchestratorAction.EXECUTE,
        next_agent=AgentType.TRANSCRIPTION,
        reason="默认执行课堂视频转写与分段流程。",
    )

File: apps/api/src/application/agent-orchestrator.js
import {
  claimNextWorkflowRun as claimWorkflowRun,
  processVideoTask,
  processWorkflowRun
} from "../../processor.js";

export const AGENT_ORCHESTRATOR_ROLES = {
  orchestrator: "decide_next_action",
  worker: "execute_background_work",
  pipeline: "run_single_processing_chain",
  integration: "connect_external_systems"
};

export const AGENT_TYPES = {
  transcription: "transcription_agent",
  translation: "translation_agent",
  evidence: "evidence_agent",
  report: "report_agent",
  teacherReview: "teacher_review_agent"
};

export async function claimNextWorkflowRun(options = {}) {
  return claimWorkflowRun(options);
}

export async function executeWorkflowRun(workflowRunId, context = {}) {
  const decision = decideWorkflowAction(context);
  if (decision.action !== "execute") return decision;
  await processWorkflowRun(workflowRunId);
  return { ...decision, status: "completed" };
}

export async function executeTask(taskId, context = {}) {
  const decision = decideWorkflowAction(context);
  if (decision.action !== "execute") return decision;
  await processVideoTask(taskId);
  return { ...decision, status: "completed" };
}

export function decideWorkflowAction({ task, workflow, userIntent } = {}) {
  if (workflow?.status === "failed" || task?.status === "failed") {
    return {
      action: "retry_or_wait_for_teacher",
      nextAgent: AGENT_TYPES.teacherReview,
      reason: "当前流程失败，需要教师确认后重试或调整配置。"
    };
  }

  if (userIntent?.requiresTranslation) {
    return {
      action: "execute",
      nextAgent: AGENT_TYPES.translation,
      reason: "教师主动要求翻译，进入可选翻译 Agent。"
    };
  }

  return {
    action: "execute",
    nextAgent: AGENT_TYPES.transcription,
    reason: "默认执行课堂视频转写与分段流程。"
  };
}
